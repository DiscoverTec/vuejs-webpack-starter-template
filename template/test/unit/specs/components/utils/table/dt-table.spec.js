import { shallow, mount } from '@vue/test-utils'
import { createRenderer } from 'vue-server-renderer'
import Component from '@/components/utils/table/dt-table.vue'

describe(`Component ${Component.name}`, () => {

  const tableColumnsSimple = [
    {
      field: 'id',
      label: 'ID'
    },
    {
      field: 'firstName',
      label: 'First Name'
    },
    {
      field: 'lastName',
      label: 'Last Name'
    },
    {
      field: 'date',
      label: 'Date'
    }
  ]

  const tableColumnsSort = [
    {
      field: 'id',
      label: 'ID',
      sortable: true
    },
    {
      field: 'firstName',
      label: 'First Name',
      sortable: true
    },
    {
      field: 'lastName',
      label: 'Last Name'
    },
    {
      field: 'date',
      label: 'Date'
    }
  ]


  const tableColumnsHidden = [
    {
      field: 'id',
      label: 'ID',
      hidden: true
    },
    {
      field: 'firstName',
      label: 'First Name',
      hidden: true
    },
    {
      field: 'lastName',
      label: 'Last Name'
    },
    {
      field: 'date',
      label: 'Date'
    }
  ]

  const tableData = [
    {
      id: 1,
      firstName: 'Jesse',
      lastName: 'Simmons',
      date: '2016/10/15 13:43:27',
    },
    {
      id: 2,
      firstName: 'John',
      lastName: 'Jacobs',
      date: '2016/12/15 06:00:53',
    },
    {
      id: 3,
      firstName: 'Tina',
      lastName: 'Gilbert',
      date: '2016/04/26 06:26:28',
    },
    {
      id: 4,
      firstName: 'Clarence',
      lastName: 'Flores',
      date: '2016/04/10 10:28:46',
    },
    {
      id: 5,
      firstName: 'Anne',
      lastName: 'Lee',
      date: '2016/12/06 14:38:38',
    }
  ]

  it('should render an empty table', () => {
    const renderer = createRenderer()
    const wrapper = shallow(Component, {})
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err)
      expect(str).toMatchSnapshot()
    })
  })

  it('should render an empty table with slot message', () => {
    const renderer = createRenderer()
    const wrapper = shallow(Component, {
      slots: {
        empty: 'This is empty message'
      }
    })
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err)
      expect(str).toMatchSnapshot()
    })
  })

  it('should render an empty table with footer message', () => {
    const renderer = createRenderer()
    const wrapper = shallow(Component, {
      slots: {
        tfoot: 'This is footer message'
      }
    })
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err)
      expect(str).toMatchSnapshot()
    })
  })

  it('should render a body with only props table data', () => {
    const renderer = createRenderer()
    const wrapper = shallow(Component, { propsData: { value: tableData } })
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err)
      expect(str).toMatchSnapshot()
    })
  })

  it('should render a header and body with only props table data and columns', () => {
    const renderer = createRenderer()
    const wrapper = shallow(Component, {
      propsData: {
        value: tableData,
        columns: tableColumnsSimple
      }
    })
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err)
      expect(str).toMatchSnapshot()
    })
  })

  // WIP Scoped Slots - https://github.com/vuejs/vue-test-utils/issues/156
  /*
  it('should render a header and body with head and body slots', () => {
    const renderer = createRenderer()
    const wrapper = mount(Component, {
      slots: {
        thead: '[H] \{{props.column.field}}',
        tbody: `
          <td> [T] \{{props.row.id}} </td>
          <td> [E] \{{props.row.firstName}} </td>
          <td> [S] \{{props.row.lastName}} </td>
          <td> [T] \{{props.row.date}} </td>
        `
      },
      propsData: {
        value: tableData,
        columns: tableColumnsSimple
      }
    })
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err)
      expect(str).toMatchSnapshot()
    })
  })

  it('should display details when expanding row', () => {
    const renderer = createRenderer()
    const wrapper = shallow(Component, {
      slots: {
        detail: 'This is Details'
      },
      propsData: {
        value: tableData,
        columns: tableColumnsSort
      }
    })
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err)
      expect(str).toMatchSnapshot()
    })
  })
  */

  it('should change sort direction on click', () => {
    const renderer = createRenderer()
    const wrapper = shallow(Component, {
      propsData: {
        value: tableData,
        columns: tableColumnsSort
      }
    })
    wrapper.findAll('th').at(0).trigger('click')
    expect(wrapper.emitted().sort).toMatchSnapshot('sort emitted')
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err)
      expect(str).toMatchSnapshot()
    })
  })

  it('should change sort direction and sort on when clicked', () => {
    const renderer = createRenderer()
    const wrapper = shallow(Component, {
      propsData: {
        value: tableData,
        columns: tableColumnsSort
      }
    })
    wrapper.findAll('th').at(1).trigger('click')
    wrapper.findAll('th').at(1).trigger('click')
    expect(wrapper.emitted().sort).toMatchSnapshot('sort emitted')
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err)
      expect(str).toMatchSnapshot()
    })
  })

  it('should not sort when column is not sortable', () => {
    const renderer = createRenderer()
    const wrapper = shallow(Component, {
      propsData: {
        value: tableData,
        columns: tableColumnsSort
      }
    })
    wrapper.findAll('th').at(2).trigger('click')
    expect(wrapper.emitted().sort).toBeFalsy()
  })

  it('should set the default sort on column and direction', () => {
    const renderer = createRenderer()
    const wrapper = shallow(Component, {
      propsData: {
        value: tableData,
        columns: tableColumnsSort,
        sortOn: 'firstName',
        sortDir: 'asc'
      }
    })
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err)
      expect(str).toMatchSnapshot()
    })
  })

  it('should omit sort on and direction if column is not sortable', () => {
    const renderer = createRenderer()
    const wrapper = shallow(Component, {
      propsData: {
        value: tableData,
        columns: tableColumnsSimple,
        sortOn: 'firstName',
        sortDir: 'asc'
      }
    })
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err)
      expect(str).toMatchSnapshot()
    })
  })

  it('should always flip sort if column is default to asc', () => {
    const renderer = createRenderer()
    const wrapper = shallow(Component, {
      propsData: {
        value: tableData,
        columns: tableColumnsSort,
        sortOn: 'firstName',
        sortDir: 'asc'
      }
    })
    wrapper.findAll('th').at(1).trigger('click')
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err)
      expect(str).toMatchSnapshot()
    })
  })

  it('should hide columns only', () => {
    const renderer = createRenderer()
    const wrapper = shallow(Component, {
      propsData: {
        columns: tableColumnsHidden
      }
    })
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err)
      expect(str).toMatchSnapshot()
    })
  })

  it('should hide columns and data when no slot is used', () => {
    const renderer = createRenderer()
    const wrapper = shallow(Component, {
      propsData: {
        value: tableData,
        columns: tableColumnsHidden
      }
    })
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err)
      expect(str).toMatchSnapshot()
    })
  })


  it('should emit toggle when toggle detail is called', () => {
    const testRow = {
      id: 1,
      isDetailVisible: false
    }
    const renderer = createRenderer()
    const wrapper = shallow(Component)
    wrapper.vm.toggleDetail(testRow)
    expect(wrapper.emitted().toggle).toMatchSnapshot()
  })
})
