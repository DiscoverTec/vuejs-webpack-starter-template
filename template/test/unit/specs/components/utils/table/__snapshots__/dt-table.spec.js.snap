// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component dt-table should always flip sort if column is default to asc 1`] = `
<table>
    <thead>
        <tr>
            <!---->
            <th class="cursor">
                ID
                <!---->
            </th>
            <th class="cursor">
                First Name
                <i aria-hidden="true" class="fa fa-arrow-circle-down"></i>
            </th>
            <th>
                Last Name
                <!---->
            </th>
            <th>
                Date
                <!---->
            </th>
        </tr>
    </thead>
    <tbody>
        <!---->
        <tr>
            <!---->
            <td> 1 </td>
            <td> Jesse </td>
            <td> Simmons </td>
            <td> 2016/10/15 13:43:27 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 2 </td>
            <td> John </td>
            <td> Jacobs </td>
            <td> 2016/12/15 06:00:53 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 3 </td>
            <td> Tina </td>
            <td> Gilbert </td>
            <td> 2016/04/26 06:26:28 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 4 </td>
            <td> Clarence </td>
            <td> Flores </td>
            <td> 2016/04/10 10:28:46 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 5 </td>
            <td> Anne </td>
            <td> Lee </td>
            <td> 2016/12/06 14:38:38 </td>
        </tr>
        <!---->
    </tbody>
    <!---->
</table>
`;

exports[`Component dt-table should change sort direction and sort on when clicked 1`] = `
<table>
    <thead>
        <tr>
            <!---->
            <th class="cursor">
                ID
                <!---->
            </th>
            <th class="cursor">
                First Name
                <i aria-hidden="true" class="fa fa-arrow-circle-down"></i>
            </th>
            <th>
                Last Name
                <!---->
            </th>
            <th>
                Date
                <!---->
            </th>
        </tr>
    </thead>
    <tbody>
        <!---->
        <tr>
            <!---->
            <td> 1 </td>
            <td> Jesse </td>
            <td> Simmons </td>
            <td> 2016/10/15 13:43:27 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 2 </td>
            <td> John </td>
            <td> Jacobs </td>
            <td> 2016/12/15 06:00:53 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 3 </td>
            <td> Tina </td>
            <td> Gilbert </td>
            <td> 2016/04/26 06:26:28 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 4 </td>
            <td> Clarence </td>
            <td> Flores </td>
            <td> 2016/04/10 10:28:46 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 5 </td>
            <td> Anne </td>
            <td> Lee </td>
            <td> 2016/12/06 14:38:38 </td>
        </tr>
        <!---->
    </tbody>
    <!---->
</table>
`;

exports[`Component dt-table should change sort direction and sort on when clicked: sort emitted 1`] = `
Array [
  Array [
    Object {
      "field": "firstName",
      "label": "First Name",
      "sortable": true,
    },
    "asc",
  ],
  Array [
    Object {
      "field": "firstName",
      "label": "First Name",
      "sortable": true,
    },
    "desc",
  ],
]
`;

exports[`Component dt-table should change sort direction on click 1`] = `
<table>
    <thead>
        <tr>
            <!---->
            <th class="cursor">
                ID
                <i aria-hidden="true" class="fa fa-arrow-circle-up"></i>
            </th>
            <th class="cursor">
                First Name
                <!---->
            </th>
            <th>
                Last Name
                <!---->
            </th>
            <th>
                Date
                <!---->
            </th>
        </tr>
    </thead>
    <tbody>
        <!---->
        <tr>
            <!---->
            <td> 1 </td>
            <td> Jesse </td>
            <td> Simmons </td>
            <td> 2016/10/15 13:43:27 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 2 </td>
            <td> John </td>
            <td> Jacobs </td>
            <td> 2016/12/15 06:00:53 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 3 </td>
            <td> Tina </td>
            <td> Gilbert </td>
            <td> 2016/04/26 06:26:28 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 4 </td>
            <td> Clarence </td>
            <td> Flores </td>
            <td> 2016/04/10 10:28:46 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 5 </td>
            <td> Anne </td>
            <td> Lee </td>
            <td> 2016/12/06 14:38:38 </td>
        </tr>
        <!---->
    </tbody>
    <!---->
</table>
`;

exports[`Component dt-table should change sort direction on click: sort emitted 1`] = `
Array [
  Array [
    Object {
      "field": "id",
      "label": "ID",
      "sortable": true,
    },
    "asc",
  ],
]
`;

exports[`Component dt-table should emit toggle when toggle detail is called 1`] = `
Array [
  Array [
    true,
    Object {
      "id": 1,
      "isDetailVisible": true,
    },
  ],
]
`;

exports[`Component dt-table should hide columns and data when no slot is used 1`] = `
<table>
    <thead>
        <tr>
            <!---->
            <th style="display:none;">
                ID
                <!---->
            </th>
            <th style="display:none;">
                First Name
                <!---->
            </th>
            <th>
                Last Name
                <!---->
            </th>
            <th>
                Date
                <!---->
            </th>
        </tr>
    </thead>
    <tbody>
        <!---->
        <tr>
            <!---->
            <td style="display:none;"> 1 </td>
            <td style="display:none;"> Jesse </td>
            <td> Simmons </td>
            <td> 2016/10/15 13:43:27 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td style="display:none;"> 2 </td>
            <td style="display:none;"> John </td>
            <td> Jacobs </td>
            <td> 2016/12/15 06:00:53 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td style="display:none;"> 3 </td>
            <td style="display:none;"> Tina </td>
            <td> Gilbert </td>
            <td> 2016/04/26 06:26:28 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td style="display:none;"> 4 </td>
            <td style="display:none;"> Clarence </td>
            <td> Flores </td>
            <td> 2016/04/10 10:28:46 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td style="display:none;"> 5 </td>
            <td style="display:none;"> Anne </td>
            <td> Lee </td>
            <td> 2016/12/06 14:38:38 </td>
        </tr>
        <!---->
    </tbody>
    <!---->
</table>
`;

exports[`Component dt-table should hide columns only 1`] = `
<table>
    <thead>
        <tr>
            <!---->
            <th style="display:none;">
                ID
                <!---->
            </th>
            <th style="display:none;">
                First Name
                <!---->
            </th>
            <th>
                Last Name
                <!---->
            </th>
            <th>
                Date
                <!---->
            </th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td colspan="999">
                <!---->
            </td>
        </tr>
    </tbody>
    <!---->
</table>
`;

exports[`Component dt-table should omit sort on and direction if column is not sortable 1`] = `
<table>
    <thead>
        <tr>
            <!---->
            <th>
                ID
                <!---->
            </th>
            <th>
                First Name
                <!---->
            </th>
            <th>
                Last Name
                <!---->
            </th>
            <th>
                Date
                <!---->
            </th>
        </tr>
    </thead>
    <tbody>
        <!---->
        <tr>
            <!---->
            <td> 1 </td>
            <td> Jesse </td>
            <td> Simmons </td>
            <td> 2016/10/15 13:43:27 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 2 </td>
            <td> John </td>
            <td> Jacobs </td>
            <td> 2016/12/15 06:00:53 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 3 </td>
            <td> Tina </td>
            <td> Gilbert </td>
            <td> 2016/04/26 06:26:28 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 4 </td>
            <td> Clarence </td>
            <td> Flores </td>
            <td> 2016/04/10 10:28:46 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 5 </td>
            <td> Anne </td>
            <td> Lee </td>
            <td> 2016/12/06 14:38:38 </td>
        </tr>
        <!---->
    </tbody>
    <!---->
</table>
`;

exports[`Component dt-table should render a body with only props table data 1`] = `
<table>
    <!---->
    <tbody>
        <!---->
        <tr>
            <!---->
            <td> { &quot;id&quot;: 1, &quot;firstName&quot;: &quot;Jesse&quot;, &quot;lastName&quot;: &quot;Simmons&quot;, &quot;date&quot;: &quot;2016/10/15 13:43:27&quot; } </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> { &quot;id&quot;: 2, &quot;firstName&quot;: &quot;John&quot;, &quot;lastName&quot;: &quot;Jacobs&quot;, &quot;date&quot;: &quot;2016/12/15 06:00:53&quot; } </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> { &quot;id&quot;: 3, &quot;firstName&quot;: &quot;Tina&quot;, &quot;lastName&quot;: &quot;Gilbert&quot;, &quot;date&quot;: &quot;2016/04/26 06:26:28&quot; } </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> { &quot;id&quot;: 4, &quot;firstName&quot;: &quot;Clarence&quot;, &quot;lastName&quot;: &quot;Flores&quot;, &quot;date&quot;: &quot;2016/04/10 10:28:46&quot; } </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> { &quot;id&quot;: 5, &quot;firstName&quot;: &quot;Anne&quot;, &quot;lastName&quot;: &quot;Lee&quot;, &quot;date&quot;: &quot;2016/12/06 14:38:38&quot; } </td>
        </tr>
        <!---->
    </tbody>
    <!---->
</table>
`;

exports[`Component dt-table should render a header and body with only props table data and columns 1`] = `
<table>
    <thead>
        <tr>
            <!---->
            <th>
                ID
                <!---->
            </th>
            <th>
                First Name
                <!---->
            </th>
            <th>
                Last Name
                <!---->
            </th>
            <th>
                Date
                <!---->
            </th>
        </tr>
    </thead>
    <tbody>
        <!---->
        <tr>
            <!---->
            <td> 1 </td>
            <td> Jesse </td>
            <td> Simmons </td>
            <td> 2016/10/15 13:43:27 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 2 </td>
            <td> John </td>
            <td> Jacobs </td>
            <td> 2016/12/15 06:00:53 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 3 </td>
            <td> Tina </td>
            <td> Gilbert </td>
            <td> 2016/04/26 06:26:28 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 4 </td>
            <td> Clarence </td>
            <td> Flores </td>
            <td> 2016/04/10 10:28:46 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 5 </td>
            <td> Anne </td>
            <td> Lee </td>
            <td> 2016/12/06 14:38:38 </td>
        </tr>
        <!---->
    </tbody>
    <!---->
</table>
`;

exports[`Component dt-table should render an empty table 1`] = `
<table>
    <!---->
    <tbody>
        <tr>
            <td colspan="999">
                <!---->
            </td>
        </tr>
    </tbody>
    <!---->
</table>
`;

exports[`Component dt-table should render an empty table with footer message 1`] = `
<table>
    <!---->
    <tbody>
        <tr>
            <td colspan="999">
                <!---->
            </td>
        </tr>
    </tbody>
    <tfoot>
        <tr>
            <td colspan="999">This is footer message</td>
        </tr>
    </tfoot>
</table>
`;

exports[`Component dt-table should render an empty table with slot message 1`] = `
<table>
    <!---->
    <tbody>
        <tr>
            <td colspan="999">This is empty message</td>
        </tr>
    </tbody>
    <!---->
</table>
`;

exports[`Component dt-table should set the default sort on column and direction 1`] = `
<table>
    <thead>
        <tr>
            <!---->
            <th class="cursor">
                ID
                <!---->
            </th>
            <th class="cursor">
                First Name
                <i aria-hidden="true" class="fa fa-arrow-circle-up"></i>
            </th>
            <th>
                Last Name
                <!---->
            </th>
            <th>
                Date
                <!---->
            </th>
        </tr>
    </thead>
    <tbody>
        <!---->
        <tr>
            <!---->
            <td> 1 </td>
            <td> Jesse </td>
            <td> Simmons </td>
            <td> 2016/10/15 13:43:27 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 2 </td>
            <td> John </td>
            <td> Jacobs </td>
            <td> 2016/12/15 06:00:53 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 3 </td>
            <td> Tina </td>
            <td> Gilbert </td>
            <td> 2016/04/26 06:26:28 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 4 </td>
            <td> Clarence </td>
            <td> Flores </td>
            <td> 2016/04/10 10:28:46 </td>
        </tr>
        <!---->
        <tr>
            <!---->
            <td> 5 </td>
            <td> Anne </td>
            <td> Lee </td>
            <td> 2016/12/06 14:38:38 </td>
        </tr>
        <!---->
    </tbody>
    <!---->
</table>
`;
